### PR TITLE
Allows any object to implement custom serialization for ActiveJob arguments

### PR DESCRIPTION
**Pull request by fabn**
I was about to open a feature request, then I saw #19418.

As told by @rafaelfranca in [this comment](https://github.com/rails/rails/issues/19418#issuecomment-83745633) currently it is possible to override `#serialize_argument` and `#deserialize_argument` but to do that a user should implement those methods in every job that accepts a custom param. 

Even using that technique it won't be possible to override some default behavior (such as passing a Date or Time object in mailers, see #18519) without monkey patching builtin classes.

The other way I found is to include `GlobalID::Identification` (#21551) and implement a custom locator but in my opinion that api should be used when argument does have some form of identifiers (e.g. a remote entity identified by some id).

For this reason I'd like to introduce a concept of custom serializer (and deserializer) in AJ. In this way library writers can provide serialization for objects defined in their library and such object may be passed as is as AJ job arguments, e.g.

``` ruby
class MyCustomType
  include ActiveJob::Serialization

  class << self

    def aj_dump(value)
      Marshal.dump(value)
    end

    def aj_load(value)
      Marshal.load(value)
    end

  end

end
# later
SomeJob.perform_later(MyCustomType.new)
```

What do you think?

P.S. in PR I'm raising `NotImplementedError` as default implementation for `aj_load` and `aj_dump`, however I'm tempted to change it with `Marshal.dump` and `Marshal.load` as default values in this way for simplest objects one could simply include the module and have everything working out of the box.

### CODE REVIEW

**Comment by fabn** [comment]
@matthewd any feedback for this?

Thanks.

**Comment by matthewd** [comment]
I believe the lack of support for general serialization is a designed feature -- see e.g. https://github.com/rails/activejob/issues/14.

r? @dhh

File: activejob/lib/active_job/serialization.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+require 'active_support/concern'
+
+module ActiveJob
+
```
Comment (sgrif) [inline activejob/lib/active_job/serialization.rb]: :scissors:

File: activejob/lib/active_job/serialization.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  # Module to be included to enable custom serialization for any object.
+  module Serialization
+
```
Comment (sgrif) [inline activejob/lib/active_job/serialization.rb]: :scissors:

File: activejob/lib/active_job/serialization.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    extend ActiveSupport::Concern
+
+    # Instance method is just a shortcut for object serialization
+    def aj_dump
+      self.class.aj_dump(self).to_s
+    end
+
+    module ClassMethods
+
```
Comment (sgrif) [inline activejob/lib/active_job/serialization.rb]: :scissors:

File: activejob/lib/active_job/serialization.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      # Returns a String that will be passed to #aj_load when executing job.
+      def aj_dump(value)
+        raise NotImplementedError, "Implementation for #{self}.#{__method__} must be provided"
+      end
+
+      # Used to deserialize an argument. Return value of this method will be actual argument passed to the job instance.
+      def aj_load(value)
+        raise NotImplementedError, "Implementation for #{self}.#{__method__} must be provided"
+      end
+
```
Comment (sgrif) [inline activejob/lib/active_job/serialization.rb]: :scissors:

File: activejob/lib/active_job/serialization.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    end
+
```
Comment (sgrif) [inline activejob/lib/active_job/serialization.rb]: :scissors:

File: activejob/lib/active_job/serialization.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
```
Comment (sgrif) [inline activejob/lib/active_job/serialization.rb]: :scissors:

**Comment by sgrif** [comment]
Would it make sense to just fall back to YAML or Marshal instead of requiring the users to do anything?

**Comment by fabn** [comment]
Yes, after pushed I made the same observation. In this case I think Marshal would be a better default.

All those scissors stands for?

**Comment by sgrif** [comment]
:scissors: == delete this line

**Comment by fabn** [comment]
Changed to use `Marshal` module as default implementation.

File: activejob/test/models/custom_type.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  # Some attributes to serialize
+  attr_reader :foo, :bar
+
+  def initialize(v1, v2)
+    @foo = v1
+    @bar = v2
+  end
+
+  # equality method used in tests
+  def ==(obj)
```
Comment (sgrif) [inline activejob/test/models/custom_type.rb]: Let's just use a Struct for this.

**Comment by rafaelfranca** [comment]
I don't think it is a good idea to use Marshal by default. It may introduce a RCE vulnerability if users can control what is going to be serialized.

File: activejob/test/models/custom_type.rb (hunk 7 above)
Comment (fabn) [inline activejob/test/models/custom_type.rb]: Do you want to keep a separated file or the whole structure inlined in test case? Something like

```
test 'should use custom serialization when implemented' do
  CustomType = Struct.new('CustomType', :foo, :bar)  do
    include ActiveJob::Serialization
  end
  value = CustomType.new(rand(1000), rand(1000))
  assert_arguments_roundtrip [value]
end
```

**Comment by fabn** [comment]
> I don't think it is a good idea to use Marshal by default. It may introduce a RCE vulnerability if users can control what is going to be serialized.

@rafaelfranca good point. Moving to YAML?

File: activejob/test/models/custom_type.rb (hunk 7 above)
Comment (sgrif) [inline activejob/test/models/custom_type.rb]: I'd inline it, but define the constant outside of the test case (still inside the test class though)

**Comment by sgrif** [comment]
I'm not sure I see the RCE vector given that if the users could control what's being serialized, they would be limited to types that could come out of `params`.

**Comment by fabn** [comment]
@rafaelfranca after thinking to this I don't believe that arguments serialization would introduce in the code a RCE vulnerability. Other libraries (e.g. [sidekiq](https://github.com/mperham/sidekiq/blob/7a727c61cfd3f174d32e3a231b14833b61b897fe/lib/sidekiq/extensions/class_methods.rb#L15-L19)) already doing this and there's no way for untrusted users to tamper serialized YAML (or Marshal data).

When custom arguments are serialized yaml only may contain untrusted input as data fields, thus unless we're doing something silly in the serialized class there's no way for user to trigger code execution.

```ruby
CustomType = Struct.new('CustomType', :foo, :bar) do
  include ActiveJob::Serialization
end
# YAML example
CustomType.new('Untrusted input', 'Something else').aj_dump
 => "--- !ruby/struct:Struct::CustomType\nfoo: Untrusted input\nbar: Something else\n"
                            user input =>      ^^^^^^^^^^^^^^^       ^^^^^^^^^^^^^^    
```

Maybe YAML instead of Marshal would be a more transparent choice here.

Another option would be to use [safe_load](http://docs.ruby-lang.org/en/2.1.0/Psych.html#method-c-safe_load) when available. In this way only basic types would be allowed and if implementors need advanced types serialization they would be forced to implement that method by themselves. Default `aj_load` implementation may become

```ruby
def aj_load(value)
  # I don't know if all ruby implementations/versions include safe_yaml, this will prevent weird errors.
  if ::YAML.respond_to?(:safe_load)
    ::YAML.self_load(value, [self])
    # maybe here we should wrap potential Psych::DisallowedClass errors in a custom exception
  else
    raise NotImplementedError, "To use automatic unserialization your YAML library must support #safe_load"
  end
end
```

In this way every chance of RCE is automatically prevented.

What do you think?

**Comment by fabn** [comment]
Added a safest default implementation for custom serialization using safe_load.

Let me know.

**Comment by rafaelfranca** [comment]
We can use `YAML.load`. There is no point of requiring a lib to use this if we want to ship this behavior by default.

**Comment by fabn** [comment]
> We can use YAML.load

Instead of `YAML.safe_load`?

**Comment by rafaelfranca** [comment]
I don't like the method names. It is hard to tell what `aj` means. We don't use abbreviations for the frameworks in any place in our documentation or code. Maybe `serialize_to_job` and `serialize_from_job` are better options?

I also don't like of having two ways to serialize an argument in activejob. Global ID and this new API. Why Global ID can't be used for these cases?

**Comment by fabn** [comment]
> I don't like the method names. It is hard to tell what aj means. We don't use abbreviations for the frameworks in any place in our documentation or code. Maybe serialize_to_job and serialize_from_job are better options?

Ok, that's fine, I can rename those methods.

> I also don't like of having two ways to serialize an argument in activejob. Global ID and this new API. Why Global ID can't be used for these cases?

From my first comment:

> The other way I found is to include GlobalID::Identification (#21551) and implement a custom locator but in my opinion that api should be used when argument does have some form of identifiers (e.g. a remote entity identified by some id).

Assume I have a simple class like the ones used in tests, or a custom form object built from a form submission (e.g. a mail form with name, email and message) and some logic encapsulated into this object. Doing `include ActiveJob::Serialization` in the class and pass it to a job instance is much more simple than implementing a `GlobalID::Identification` (or than serialize it into a hash, pass the hash to job and restore in job instance).

I had the idea when reading [this issue](https://github.com/plataformatec/mail_form/issues/36#issuecomment-91940732). That good gem can't use `deliver_later` as is because their form object cannot be serialized and passed to AJ.

I think this proposal covers a different use case from GlobalID.

In any case you are free to tell me, I don't see the point and close the PR :wink:

**Comment by rafaelfranca** [comment]
I think it is possible to use `GlobalID::Identification` to serialize the Form object, you only need to implement it. I really don't like the idea of having two ways to serialize things to Active Job.

Adding this API I don't see a reason why I could not just remove GlobalID and serialize Active Record objects using YAML or Marshal.

I'm fine with `ActiveJob::Serialization` being a default implementation to serialize object using Global ID, can you try to go in that direction?

**Comment by fabn** [comment]
I understand your point, but as I said before I think they are two completely different things. Quoting Global ID project page

> A Global ID is an app wide URI that uniquely identifies a model instance:

Global ID are meant to be used when a (potentially huge) object is uniquely identified by a short identifier that not carry with it the whole object. Not all objects have this property and for this reason they cannot be represented in this way.

Object serialization, on the contrary, carry with it the whole object with all its properties and all data needed to reconstruct the object as is before serialization.

Also when creating URIs with serialization they can became pretty long and useless as URI, e.g.

```
# Possible output for a Struct.new('CustomType', :foo, :bar) instance converted to a global id
gid://aj/Struct::CustomType/---+%21ruby%2Fstruct%3AStruct%3A%3ACustomType%0Afoo%3A+81%0Abar%3A+607%0A
```

Imagine the GID serialization for an email message body.

That said this would be the quickest way to integrate global ID with `ActiveJob::Serialization`. I don't like it at all. Overriding `.find` and `#id`for a generic class imho is a very bad idea.

```patch
diff --git a/activejob/lib/active_job/serialization.rb b/activejob/lib/active_job/serialization.rb
index 5989388..0d18587 100644
--- a/activejob/lib/active_job/serialization.rb
+++ b/activejob/lib/active_job/serialization.rb
@@ -14,11 +14,20 @@ module ActiveJob
   module Serialization
     extend ActiveSupport::Concern
 
+    included do
+      include GlobalID::Identification
+    end
+
     # Instance method is just a shortcut for object serialization
     def serialize_to_job
       self.class.serialize_to_job(self).to_s
     end
 
+    # Test identifier
+    def id
+      serialize_to_job
+    end
+
     module ClassMethods
       # Return a serialized version of the argument. It must return a `#to_s` compatible object that will be
       # passed to `#load` when unserialized.
@@ -44,6 +53,11 @@ module ActiveJob
           raise NotImplementedError, 'To use automatic unserialization your YAML library must support #safe_load'
         end
       end
+
+      # Test finder
+      def find(id)
+        serialize_from_job(id)
+      end
     end
   end
 end
```

Another possibility to integrate would be to use some kind of key value app storage (Rails.cache ??), store object serialization in that storage, generate a UUID for the object and return that UUID as gid.

When locating the object the process would be pretty simple, just lookup in the storage and deserialize.

This would fit better your idea (far better than the patch above) however Rails.cache is not so appropriate because it's volatile and I don't see other options than relying on other dependencies.

**Comment by fabn** [comment]
@rafaelfranca, @sgrif comments?

**Pull request closed without merging**